streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1985,6 +1989,30 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
     }
   }
 
+  /**
+   * Ensures that the cleanup function runs, whether this stream succeeds, fails,
+   * or is interrupted, after this stream's finalizers run.
+   */
+  final def onExit[R1 <: R](cleanup: Exit[E, Any] => ZIO[R1, Nothing, Any]): ZStream[R1, E, A] =
```

Comment:
In the current formulation, this needs to be `Exit[Any, Unit]`. The reason is that `ZManaged.finalizerRef` is actually fed the `Exit` value resulting from `ZManaged#use`. `use` is universally quantified on `A`, therefore you've correctly set `A = Any` here. However, the `E` in the `Exit` is `E1 >: E` - meaning it is any arbitrary supertype of `E`, which means you cannot safely cast it to `E` as is done below.

Another formulation exists: you can modify the returned `Pull` from `self.process` to store the `Exit` in a `Ref[Exit[E, A]]`. Then you can safely feed that to the cleanup function. What we'd be doing here is basically storing the last pulled value every time `self` is pulled.

Reply:
1. I'm not sold that the success branch can have `Unit`, indeed because of the fact that it is universally quantified, which means I'd expect it to be `Any`. But check 3.
2. For the error type, you are correct, I incorrectly assumed that the only error that can happen is the stream pull error `E`.
3. Regarding the other formulation, I thought about it, and I see it like this. If the stream always succeeds and signals success on each pull, the only relevant `Exit` value is the last one, which is the `Exit` value for `Pull.end` (and for example, we explicitly signal this with `Unit` in `flatMap`, so maybe you are right, we should be more explicit about that in the Exit type). On the flip side, if there is an unhandled error, the scope exiting will start to unwind and finish with the `Any` type of the error, which is again, the only relevant `Exit` value. So this tells me that keeping and updating the `Ref[Exit[E, A]]` on each pull is redundant.

So my proposal is the following, we go either with an `Exit[Any, Any]` and give no expectation for the types of error or success, or go with `Exit[Any, Unit]` as you mentioned but manually succeed with a `Unit` in these combinators and anywhere else that is needed to manually run finalizers (e.g. `flatMap`). Honestly, I'd go with the second one if we do decide to eventually go the new encoding way of finishing with a value `B`, which we can switch to in the future. WDYT?

Reply:
I didn't quite follow what you were saying with 3. I will simplify what I'm saying. There are two possible interpretations for this combinator:
- The cleanup is provided with the `Exit` of processing the stream - that is, the result of `foreach/run`. This is `Exit[Any, Any]` and can be written with the current formulation.
- The cleanup is provided with the `Exit` of the last pull from the stream. This is `Exit[E, A]`, and requires the Ref.

Which one is more useful?